File: GhostBot/controller/threaded_bot_controller.py
import threading
import time
from typing import Callable

from GhostBot.client_launcher import ClientLauncher
from GhostBot.controller.bot_controller import BotController, BotClientWindow
from GhostBot.controller.login_controller import LoginController
from GhostBot.enums.bot_status import BotStatus
from GhostBot.lib.win32.process import PymemProcess


class ThreadedBotController(BotController):
    def __init__(self, auto_login: bool = True, **kwargs):
        super().__init__(**kwargs)
        self._tasks: dict[str, threading.Thread] = dict()
        self.auto_login = auto_login

    def scan_for_clients(self):
        try:
            while self._running:
                self.logger.debug("uptime %ss", self._total_running_secs)
                self._remove_closed_pending_clients()
                self._scan_for_clients()

                if self.auto_login:
                    self._process_login_queue()

                for i in range(0, 10):
                    if not self._running:
                        break
                    time.sleep(1)
        except KeyboardInterrupt:
            self.logger.info('Exiting as requested...')
            return
        except Exception as e:
            self.logger.exception(e)
            return

    def _process_login_queue(self):

        def callback(__client, result):
            if not result:
                self.logger.debug('[%s] Login failed, removing %s from self._pending_clients', __client.process_id, __client.name)
                try:
                    self._pending_clients.pop(__client.name)
                except KeyError:
                    self.logger.debug('[%s] pending client not found.. possibly harmless', __client.name)
            else:
                self.logger.debug('[%s] :: Login succeeded', __client.process_id)
                self._pending_clients.pop(__client.name)
                self.add_client(__client)

        eligible_logins = self._eligible_logins()

        if not eligible_logins:
            self.logger.debug('no eligible logins...')
            return

        elif len(self.login_queue) < 1:
            try:
                if len(list(PymemProcess.get_proc_matching(b'game.exe'))) == 0:
                    self.logger.debug('spawning game launcher')
                    ClientLauncher().block_until_ready()
                else:
                    self.logger.debug('game launcher already running')
            except IndexError:
                self.logger.debug('game launcher process didnt launch, retrying')
            except KeyError:
                self.logger.debug('too many game launcher processes detected, there can only be one')

        for pid, _client in dict(self.login_queue).items():

            if f"task{pid}" not in self._tasks.keys():
                lc = LoginController(_client, self)

                if lc.current_stage == LoginController.LoginStage.character_select:
                    time.sleep(5)  # fixes a race condition with the client window opening

                __char = _client.get_window_name()

                self.logger.debug('self._eligible_logins :: %s', eligible_logins)

                if __char in eligible_logins and _client.name is None:
                    _conf = eligible_logins.pop(__char)
                    self.logger.info('[%s|%s] resuming login procedure with config %s', pid, _conf.char_name, _conf)
                elif lc.current_stage == LoginController.LoginStage.enter_credentials:
                    __char, _conf = eligible_logins.popitem()
                    self.logger.info('[%s|%s] starting login procedure with config %s', pid, _conf.char_name, _conf)
                else:
                    self.logger.info('[%s|%s] skipping', pid, __char)
                    continue

                if __char in self.requested_logins:
                    self.requested_logins.remove(__char)

                self._pending_clients[_conf.char_name] = _client

                self.logger.debug('LoginStage of new pending client :: %s', lc.current_stage)

                # if lc.current_stage == LoginController.LoginStage.enter_credentials:
                self.logger.debug('[%s|%s] setting config for LoginController', pid, _conf.char_name)
                lc.set_config(_conf)
                self._add_task(lc.handle_login, f"task{pid}", lambda result: callback(_client, result))
                del self.login_queue[pid]

    def _remove_closed_pending_clients(self):
        current_running_procs = self._pymem_process.list_clients()
        for k, v in list(self._pending_clients.items()):
            if (c_pid := v.proc.process_id) not in (p.process_id for p in current_running_procs):
                self.logger.info("BotController :: removing [%s]", c_pid)
                try:
                    self._stop_task(f'task{self._pending_clients.pop(k).process_id}')
                except Exception as e:
                    self.logger.info(e)

    def _add_task(self, target: Callable, task_name: str, *target_args, **target_kwargs):
        task = threading.Thread(target=target, args=target_args, kwargs=target_kwargs)
        self._tasks[task_name] = task
        task.start()

    def _stop_task(self, task_name: str, timeout: int = 30):
        self._tasks.get(task_name).join(timeout=timeout)

    def _stop_all_tasks(self, timeout: int = 30):
        for _task_name, _task in self._tasks.items():
            self.logger.info('stopping task %s', _task_name)
            _task.join(timeout=timeout)

    def start_bot(self, client: BotClientWindow | str) -> None:
        if isinstance(client, str):
            _client = client
            if (client := self.get_client(client)) is None:
                self.logger.warning('no client %s', _client)
                return
        self.reload_bot_config(client)
        client.start_bot()
        self._add_task(self._run_bot, client.name, client)

    def _run_bot(self, client: BotClientWindow) -> None:
        client.bot_status = BotStatus.started
        self.logger.info("%s: Started.", client.name)

        functions = list(self._get_functions_for_client(client))
        for function in functions:
            if self._ipc_log_handler not in function.logger.handlers:
                function.logger.addHandler(self._ipc_log_handler)

        while client.running:
            client.bot_status = BotStatus.running
            if client.disconnected:
                self.logger.info("%s: disconnected.", client.name)
                client.close_window()
                break
            try:
                for function in functions:
                    function.run()

            except Exception as e:
                self.logger.exception(e)
        client.bot_status = BotStatus.stopped
        self.logger.info("%s: Stopped.", client.name)

    def stop_all_bots(self, timeout=30) -> None:
        self.logger.info("stopping all bots...")
        _stopping = []
        for client in self.clients.values():
            if client.running:
                self.logger.info('stopping client  %s', client.name)
                client.stop_bot()
                _stopping.append(client)
        for client in _stopping:
            self.logger.debug('joining thread %s', client.name)
            self._stop_task(client.name, timeout)
            client.bot_status = BotStatus.stopped

    def stop_bot(self, client: str | BotClientWindow, timeout=5) -> None:
        if isinstance(client, str):
            client: BotClientWindow = self.get_client(client)
            if client is None:
                self.logger.warning('no client %s', client)
                return

        if client.running:
            client.stop_bot()
            self._stop_task(client.name, timeout)
            client.bot_status = BotStatus.stopped

    def listen(self):
        self._running = True
        try:
            self._controller_start_time = time.time()
            self._add_task(self.scan_for_clients, "scan_for_clients")
            self.server.run()
            while self._running:
                time.sleep(1)
        except KeyboardInterrupt:
            self.logger.info('Exiting...')
        except Exception as e:
            self.logger.exception(e)
        self.shutdown()

    def shutdown(self):
        super().shutdown()
        self._stop_all_tasks()

if __name__ == '__main__':
    import os
    import logging
    from GhostBot import logger as _logger

    if os.environ.get('PYCHARM_HOSTED'):
        _logger.setLevel(logging.DEBUG)

    bot_controller = ThreadedBotController()
    try:
        bot_controller.listen()
    except KeyboardInterrupt:
        _logger.info('received signal, exiting')
    finally:
        _logger.info('exiting...')
        bot_controller.shutdown()

File: kotlin-server/src/main/kotlin/com/ghostbot/controller/ThreadedBotController.kt
package com.ghostbot.controller

import com.ghostbot.ClientLauncher
import com.ghostbot.config.LoginDetailsConfigLoader
import com.ghostbot.enums.BotStatus
import com.ghostbot.lib.sleepSeconds
import com.ghostbot.win32.ProcessMemory
import java.util.logging.Level

/**
 * Port of GhostBot/controller/threaded_bot_controller.py.
 */
class ThreadedBotController(
    val autoLogin: Boolean = true,
    host: String? = null,
    port: Int? = null,
    closeDisconnectedClients: Boolean = true,
) : BotController(host, port, closeDisconnectedClients) {

    private val tasks = LinkedHashMap<String, Thread>()

    override fun scanForClients() {
        try {
            while (_running) {
                  logger.log(Level.FINE, "uptime ${totalRunningSecs}s")
                try {
                    removeClosedPendingClients()
                    super.scanForClients()

                    if (autoLogin) {
                        processLoginQueue()
                    }
                } catch (e: java.lang.UnsatisfiedLinkError) {
                    // Win32 process enumeration (kernel32/user32 via JNA) only
                    // exists on Windows. An UnsatisfiedLinkError is an Error, not
                    // an Exception, so it would otherwise kill this thread
                    // silently. Stop the scan loop; the IPC server keeps serving.
                    logger.log(
                        Level.SEVERE,
                        "Win32 process enumeration is not available on this OS " +
                            "(${e.message}); scan loop disabled, IPC server continues."
                    )
                    break
                }

                for (i in 0 until 10) {
                    if (!_running) break
                    sleepSeconds(1.0)
                }
            }
        } catch (e: InterruptedException) {
            logger.info("Exiting as requested...")
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "scan loop error", e)
        }
    }

    private fun processLoginQueue() {
        val callback = { client: BotClientWindow, result: Boolean ->
            if (!result) {
                logger.log(
                    Level.FINE,
                    "[${client.processId}] Login failed, removing ${client.name} from self._pending_clients"
                )
                runCatching { pendingClients.remove(client.name) }
                    .onFailure {
                        logger.log(Level.FINE, "[${client.name}] pending client not found.. possibly harmless")
                    }
            } else {
                logger.log(Level.FINE, "[${client.processId}] :: Login succeeded")
                pendingClients.remove(client.name)
                addClient(client)
            }
        }

        val eligibleLogins = eligibleLogins()

        if (eligibleLogins.isEmpty()) {
            logger.log(Level.FINE, "no eligible logins...")
            return
        } else if (loginQueue.size < 1) {
            try {
                if (ProcessMemory.getProcMatching("game.exe").isEmpty()) {
                    logger.log(Level.FINE, "spawning game launcher")
                    ClientLauncher().blockUntilReady()
                } else {
                    logger.log(Level.FINE, "game launcher already running")
                }
            } catch (e: NoSuchElementException) {
                logger.log(Level.FINE, "game launcher process didnt launch, retrying")
            } catch (e: Exception) {
                logger.log(Level.FINE, "too many game launcher processes detected, there can only be one")
            }
        }

        for ((pid, client) in LinkedHashMap(loginQueue)) {
            if ("task$pid" in tasks.keys) continue

            val lc = LoginController(client, this)

            if (lc.currentStage == LoginController.LoginStage.CHARACTER_SELECT) {
                // fixes a race condition with the client window opening
                sleepSeconds(5.0)
            }
            val windowChar = client.getWindowName()

            logger.log(Level.FINE, "self._eligible_logins :: $eligibleLogins")

            var conf: LoginDetailsConfigLoader.CharDetails? = null
            when {
                windowChar in eligibleLogins && client.name == null -> {
                    conf = eligibleLogins.remove(windowChar)
                    logger.info("[$pid|${conf?.charName}] resuming login procedure with config $conf")
                }
                lc.currentStage == LoginController.LoginStage.ENTER_CREDENTIALS -> {
                    if (eligibleLogins.isEmpty()) {
                        logger.warning("[$pid|$windowChar] no eligible logins left for enter_credentials stage, skipping")
                        continue
                    }
                    val (key, value) = eligibleLogins.entries.first()
                    eligibleLogins.remove(key)
                    conf = value
                    logger.info("[$pid|${conf.charName}] starting login procedure with config $conf")
                }
                else -> {
                    logger.info("[$pid|$windowChar] skipping")
                    continue
                }
            }
            val selected = conf!!
            if (selected.charName in requestedLogins) {
                requestedLogins.remove(selected.charName)
            }

            pendingClients[selected.charName] = client

            logger.log(Level.FINE, "LoginStage of new pending client :: ${lc.currentStage}")

            logger.log(Level.FINE, "[$pid|${selected.charName}] setting config for LoginController")
            lc.setConfig(selected)
            addTask({ lc.handleLogin { result -> callback(client, result) } }, "task$pid")
            loginQueue.remove(pid)
        }
    }

    private fun removeClosedPendingClients() {
        val currentRunningProcs = ProcessMemory.listClients()
        for ((k, v) in LinkedHashMap(pendingClients)) {
            val cPid = v.processId
            if (cPid !in currentRunningProcs.map { it.processId }) {
                logger.info("BotController :: removing [$cPid]")
                try {
                    stopTask("task${pendingClients.remove(k)?.processId}")
                } catch (e: Exception) {
                    logger.info(e.toString())
                }
            }
        }
    }

    private fun addTask(target: () -> Unit, taskName: String) {
        val task = Thread(target, taskName)
        task.isDaemon = true
        tasks[taskName] = task
        task.start()
    }

    private fun stopTask(taskName: String, timeout: Int = 30) {
        tasks[taskName]?.join(timeout * 1000L)
    }

    private fun stopAllTasks(timeout: Int = 30) {
        for ((taskName, task) in tasks) {
            logger.info("stopping task $taskName")
            task.join(timeout * 1000L)
        }
    }

    override fun startBot(client: BotClientWindow) {
        reloadBotConfig(client)
        client.startBot()
        addTask({ runBot(client) }, client.name ?: "unknown")
    }

    private fun runBot(client: BotClientWindow) {
        client.botStatus = BotStatus.STARTED
        logger.info("${client.name}: Started.")

        val functions = getFunctionsForClient(client)
        for (function in functions) {
            if (function.logger.handlers.none { it === ipcLogHandler }) {
                function.logger.addHandler(ipcLogHandler)
            }
        }

        while (client.running) {
            client.botStatus = BotStatus.RUNNING
            if (client.disconnected) {
                logger.info("${client.name}: disconnected.")
                client.closeWindow()
                break
            }
            try {
                for (function in functions) {
                    function.run()
                }
            } catch (e: Exception) {
                logger.log(Level.SEVERE, "bot loop error for ${client.name}", e)
            }
        }
        client.botStatus = BotStatus.STOPPED
        logger.info("${client.name}: Stopped.")
    }

    override fun stopAllBots(timeout: Int) {
        logger.info("stopping all bots...")
        val stopping = mutableListOf<BotClientWindow>()
        for (client in clients.values.toList()) {
            if (client.running) {
                logger.info("stopping client  ${client.name}")
                client.stopBot()
                stopping.add(client)
            }
        }
        for (client in stopping) {
            logger.log(Level.FINE, "joining thread ${client.name}")
            stopTask(client.name ?: "unknown", timeout)
            client.botStatus = BotStatus.STOPPED
        }
    }

    override fun stopBot(client: BotClientWindow, timeout: Int) {
        if (client.running) {
            client.stopBot()
            stopTask(client.name ?: "unknown", timeout)
            client.botStatus = BotStatus.STOPPED
        }
    }

    override fun listen() {
        _running = true
        try {
            controllerStartTime = System.currentTimeMillis()
            addTask({ scanForClients() }, "scan_for_clients")
            server.run()
            while (_running) {
                sleepSeconds(1.0)
            }
        } catch (e: InterruptedException) {
            logger.info("Exiting...")
        } catch (e: Exception) {
            logger.log(Level.SEVERE, "listen error", e)
        }
        shutdown()
    }

    override fun shutdown() {
        super.shutdown()
        stopAllTasks()
    }
}
